Tests/PostgresNIOTests/PostgresNIOTests.swift

```diff
@@ -348,6 +348,38 @@ final class PostgresNIOTests: XCTestCase {
         }
     }
     
+    /// https://github.com/vapor/nio-postgres/issues/27
+    func testTextDates() throws {
+        let conn = try PostgresConnection.test(on: eventLoop).wait()
+        defer { try! conn.close().wait() }
+        struct Dates: Decodable {
+            var date: Date
+            var timestamp: Date
+            var timestamptz: Date
+        }
+        let results = try conn.simpleQuery("""
+        SELECT
+            '2016-01-18'             as date,
```

Comment:
I think you are just selecting strings here. To cast in postgres, you use `::`. For example:

```sql
SELECT '2016-01-18'::DATE as date
```

Maybe that will reduce the number of permutations we need to support here?

Reply:
Here are actual values from setting a breakpoint on the .text case within PostgresData+Date.swift:

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2020-07-15"

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2012-01-02 12:34:56"

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2020-07-15 10:26:41.426138"

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2012-01-02 12:34:56-05"

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2020-07-15 10:26:41.426138-04"

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2012-01-02 23:19:56+05:45"

(lldb) po value.readString(length: value.readableBytes)
▿ Optional<String>
  - some : "2020-07-15 20:11:41.426138+05:45"

As I have noted on several occasions, there are (at least) seven variants produced by PostgreSQL 12.  These are from a database whose columns have types 'date' (first one), 'timestamp' (next two), and 'timestamptz' (final four).  For simplicity, I just used strings in the unit test.

Reply:
Ah okay. So those do indeed come back from DATE / TIMESTAMP / TIMESTAMPTZ columns. Fair enough.

What about attempting to parse them with a simpler algorithm, something like:

```swift
let timestamp = "2020-07-15 10:26:41.426138-04"
let parts = timestamp.split(separator: " ")
switch parts.count {
case 1:
    // just parse date
    // like 2020-07-15
    let date = parts[0]
case 2:
    // parse date and time
    // like 2012-01-02 12:34:56
    // like 2020-07-15 10:26:41.426138
    // like 2012-01-02 23:19:56+05:45
    let date = parts[0]
    let time = parts[1]
    if time.contains(".") {
        // parse time with . component
    } else if time.contains("+") {
        // parse time with + component
    } else {
        // parse simple time 
    }
default: fatalError("invalid")
}
```

These checks should perform a lot faster than regex and we can use them as a branching mechanism to determine which date formatter to use.

Reply:
I did consider something along those lines, i.e. using a simple classifier and then passing the results to the appropriate DateFormatter.  I decided not to implement that approach for two reasons.  First, managing five or six DateFormatters to cover the different options looked like it would add considerable complexity.  Second, and more importantly, this did not seem to address the situation where an eight format was used by Postgres.  If this happened, there was a possibility that the _wrong_ result would be returned.

I went with the regular expression because it allowed me to exactly cover the seven formats of which I am aware.  I was able to extract the digits at the same time and pass the result to the Apple library function DateComponents to do the conversion.  If the event of an unexpected eighth format, the regex will return nil allowing the error to be more easily detected.

I recognize that other authors might solve the same problem with different approaches.  I fully understand if there is a preference to avoid regular expressions.  That stated, let's look at the regular expression that I'm using.  There are three components:

Date: (?<year>[0-9]{4})-(?<month>[0-9]{2})-(?<day>[0-9]{2})
Time: (?<hour>[0-9]{2}):(?<min>[0-9]{2}):(?<sec>[0-9]{2})(?<micro>\\.[0-9]{1,6})?
TZ: (?<tzhr>[-+][0-9]{1,4})(?:[:](?<tzmin>[0-9]{2}))?

that enforce the format for each component.  These are combined in the expression:

^<Date>(?: <Time>(?:<TZ>)?)?

Here the incoming string must start with a date and can be followed by an optional time.  Only if the time is present can an optional timezone appear.

Reply:
Thanks for the explanation. That sees fair to me, then. I'm good with moving forward on this. Would you mind updating your branch to the latest master? That might help fix the CI failure.